File: src/main/java/com/matamercer/microblog/security/CustomAuthenticationFailureHandler.java
package com.matamercer.microblog.security;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.MessageSource;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.WebAttributes;
import org.springframework.security.web.authentication.SimpleUrlAuthenticationFailureHandler;
import org.springframework.web.servlet.LocaleResolver;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Locale;

public class CustomAuthenticationFailureHandler extends SimpleUrlAuthenticationFailureHandler {
    @Autowired
    private MessageSource messageSource;

    @Autowired
    private LocaleResolver localeResolver;

    @Override
    public void onAuthenticationFailure(HttpServletRequest request,
                                        HttpServletResponse response, AuthenticationException exception)
            throws IOException, ServletException {
        setDefaultFailureUrl("/login.html?error=true");

        super.onAuthenticationFailure(request, response, exception);

        Locale locale = localeResolver.resolveLocale(request);

        String errorMessage = messageSource.getMessage("message.badCredentials", null, locale);

        if (exception.getMessage().equalsIgnoreCase("User is disabled")) {
            errorMessage = messageSource.getMessage("auth.message.disabled", null, locale);
        } else if (exception.getMessage().equalsIgnoreCase("User account has expired")) {
            errorMessage =messageSource.getMessage("auth.message.expired", null, locale);
        }

        request.getSession().setAttribute(WebAttributes.AUTHENTICATION_EXCEPTION, errorMessage);
    }
}

File: src/main/java/com/matamercer/microblog/security/CustomAuthenticationFailureHandler.kt
package com.matamercer.microblog.security

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.MessageSource
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.WebAttributes
import org.springframework.security.web.authentication.SimpleUrlAuthenticationFailureHandler
import org.springframework.stereotype.Component
import org.springframework.web.servlet.LocaleResolver
import java.io.IOException
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component("authenticationFailureHandler")
class CustomAuthenticationFailureHandler @Autowired constructor(
    private val messageSource: MessageSource,
    private val localeResolver: LocaleResolver
) : SimpleUrlAuthenticationFailureHandler() {
    @Throws(IOException::class, ServletException::class)
    override fun onAuthenticationFailure(
        request: HttpServletRequest,
        response: HttpServletResponse,
        exception: AuthenticationException
    ) {
        setDefaultFailureUrl("/login?error=true")
        super.onAuthenticationFailure(request, response, exception)
        val locale = localeResolver.resolveLocale(request)
        var errorMessage = messageSource.getMessage("message.badCredentials", null, locale)
        if (exception.message.equals("User is disabled", ignoreCase = true)) {
            errorMessage = messageSource.getMessage("auth.message.disabled", null, locale)
        } else if (exception.message.equals("User account has expired", ignoreCase = true)) {
            errorMessage = messageSource.getMessage("auth.message.expired", null, locale)
        } else if (exception.message.equals("blocked", ignoreCase = true)) {
            errorMessage = messageSource.getMessage("auth.message.blocked", null, locale)
        } else if (exception.message.equals("unusual location", ignoreCase = true)) {
            errorMessage = messageSource.getMessage("auth.message.unusual.location", null, locale)
        }
        request.session
            .setAttribute(WebAttributes.AUTHENTICATION_EXCEPTION, errorMessage)
    }
}